encrypt_util.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2021 Snowflake Computing Inc. All right reserved.
+
+package gosnowflake
+
+//lint:file-ignore U1000 TODO SNOW-29352
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"strconv"
+)
+
+type snowflakeFileEncryption struct {
+	QueryStageMasterKey string `json:"queryStageMasterKey,omitempty"`
+	QueryID             string `json:"queryId,omitempty"`
+	SMKID               int64  `json:"smkId,omitempty"`
+}
+
+func (sfe *snowflakeFileEncryption) encryptStream(src *bytes.Buffer, out *bytes.Buffer, chunkSize int) encryptMetadata {
+	if chunkSize == 0 {
+		chunkSize = aes.BlockSize * 4 * 1024
+	}
+	decodedKey, _ := base64.StdEncoding.DecodeString(sfe.QueryStageMasterKey)
+	keySize := len(decodedKey)
+
+	em := new(encryptMetadata)
+	ivData := em.getSecureRandom(aes.BlockSize)
+	fileKey := em.getSecureRandom(keySize)
+
+	block, err := aes.NewCipher(fileKey)
+	if err != nil {
+		panic(err)
+	}
+	mode := cipher.NewCBCEncrypter(block, ivData)
+	cipherText := make([]byte, aes.BlockSize+src.Len()) // sizeOf(initializationVector) + sizeOf(src)
+
+	padded := false
+	for {
+		buf := getByteBufferContent(&src)
+		if len(*buf) == 0 {
+			break
+		} else if len(*buf)&aes.BlockSize != 0 {
```

Comment:
Should this be `%` and not `&` ?

Reply:
If it _should_ be `%` then this logic would mean that you pad ONLY if you have residual bytes or only if srcLen is not a multiple of aes.BlockSize whereas you should be padding either ways. In case the srcLen is a multiple of block size you should add one whole block of padding.

Reply:
You are correct. We are indeed padding only if we have residual bytes that is not a multiple of aes.BlockSize, but why would we have to pad if the stream length is a multiple of the aes.BlockSize?

Reply:
Consider the case when the last byte is 0x01 do you consider it a padding or part of the data?